lib/parachain/validation_protocol.go

```diff
@@ -8,23 +8,199 @@ import (
 
 	"github.com/ChainSafe/gossamer/dot/network"
 	"github.com/ChainSafe/gossamer/lib/common"
+	parachaintypes "github.com/ChainSafe/gossamer/lib/parachain/types"
 	"github.com/ChainSafe/gossamer/pkg/scale"
 	"github.com/libp2p/go-libp2p/core/peer"
 )
 
 const MaxValidationMessageSize uint64 = 100 * 1024
 
-type ValidationProtocolV1 struct {
-	//	TODO: Implement this struct https://github.com/ChainSafe/gossamer/issues/3318
+type UncheckedSignedAvailabilityBitfield struct {
+	// The payload is part of the signed data. The rest is the signing context,
+	// which is known both at signing and at validation.
+	Payload scale.BitVec `scale:"1"`
+
+	// The index of the validator signing this statement.
+	ValidatorIndex parachaintypes.ValidatorIndex `scale:"2"`
+
+	// The signature by the validator of the signed payload.
+	Signature ValidatorSignature `scale:"3"`
+}
+
+type Bitfield struct {
+	Hash                                common.Hash                         `scale:"1"`
+	UncheckedSignedAvailabilityBitfield UncheckedSignedAvailabilityBitfield `scale:"2"`
+}
+
+// Index returns the VaryingDataType Index
+func (Bitfield) Index() uint {
+	return 0
+}
+
+type BitfieldDistributionMessage scale.VaryingDataType
+
+func NewBitfieldDistributionMessageVDT() BitfieldDistributionMessage {
+	vdt := scale.MustNewVaryingDataType(Bitfield{})
+	return BitfieldDistributionMessage(vdt)
+}
+
+func (BitfieldDistributionMessage) New() BitfieldDistributionMessage {
+	return NewBitfieldDistributionMessageVDT()
+}
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (bdm *BitfieldDistributionMessage) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*bdm)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+
+	*bdm = BitfieldDistributionMessage(vdt)
+	return nil
+}
+
+// Value returns the value from the underlying VaryingDataType
+func (bdm *BitfieldDistributionMessage) Value() (scale.VaryingDataTypeValue, error) {
+	vdt := scale.VaryingDataType(*bdm)
+	return vdt.Value()
+}
+
+type BitfieldDistribution BitfieldDistributionMessage
+
+func NewBitfieldDistributionVDT() BitfieldDistribution {
+	vdt := scale.MustNewVaryingDataType(Bitfield{})
+	return BitfieldDistribution(vdt)
+}
+
+func (BitfieldDistribution) Index() uint {
+	return 1
+}
+func (BitfieldDistribution) New() BitfieldDistribution {
+	return NewBitfieldDistributionVDT()
+}
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (bd *BitfieldDistribution) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*bd)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+
+	*bd = BitfieldDistribution(vdt)
+	return nil
+}
```

Comment:
Can you check if just adding `Index` method here is sufficient? These creates a lot of repetitive methods.

Reply:
Removed unneeded functions.